Support Dynamic Collection Initializers

333fred opened the pull request:
As part of support DynamicInvocations in collection initializers, I
fixed a couple of bugs in C#:
1) Nested collection initializers with a missing Set method would
generate an IInvalidOperation when there was nothing invalid in the
tree.
2) DynamicMemberReferences weren't being generated for dynamic
collection Add calls, so you couldn't tell what method was getting
called by the invocation.

Addressed a small followup from #27126.

Tag @dotnet/roslyn-compiler @AlekseyTs @dotnet/analyzer-ioperation for review.

333fred commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -664,7 +664,9 @@ private IOperation CreateBoundObjectInitializerMemberOperation(BoundObjectInitia
-                        var accessor = property.GetOwnOrInheritedSetMethod();
+                        // In collection initializers, the property is not actually set. Instead, it is retrieved for a series of Add methods, so we need
+                        // to use the getter
+                        MethodSymbol accessor = inCollectionInitializer ? property.GetOwnOrInheritedGetMethod() : property.GetOwnOrInheritedSetMethod();
```
Nested collection initializers use the Get method, as they're calling Add on the collection, rather than setting the field. If the setter was missing, it would generate an InvalidOperation.

333fred commented on src/Compilers/VisualBasic/Portable/Binding/Binder_Latebound.vb:
```diff
@@ -70,10 +70,6 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
             Dim result = New BoundLateMemberAccess(node, name, containerType, receiver, boundTypeArguments, LateBoundAccessKind.Unknown, objType)
 
-            If receiver IsNot Nothing AndAlso Not receiver.WasCompilerGenerated AndAlso result.Syntax Is receiver.Syntax Then
```
Followup from https://github.com/dotnet/roslyn/pull/27126#issuecomment-392873966. IOperation test hook passed locally on all VB compiler tests.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -664,7 +664,9 @@ private IOperation CreateBoundObjectInitializerMemberOperation(BoundObjectInitia
                     else
                     {
-                        var accessor = property.GetOwnOrInheritedSetMethod();
+                        // In collection initializers, the property is not actually set. Instead, it is retrieved for a series of Add methods, so we need
```
>// In collection initializers, the property is not actually set. Instead, it is retrieved for a series of Add methods, so we need [](start = 24, length = 129)

I think this should also apply to the case when a property is initialized with a nested initializer. Is this scenario handled properly?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -765,7 +767,15 @@ private IDynamicInvocationOperation CreateBoundDynamicCollectionElementInitializ
+                                                                                                               typeArgumentsOpt: ImmutableArray<TypeSymbol>.Empty,
+                                                                                                               memberName: "Add",
+                                                                                                               implicitReceiver.Syntax,
+                                                                                                               implicitReceiver.Type,
```
>implicitReceiver.Type [](start = 111, length = 21)

I think type should be null here, it is definitely not the same as the type of the receiver

333fred replied:
For consistency with other DynamicMemberReferences, I'll make this the DynamicType

---
In reply to: [192265008](https://github.com/dotnet/roslyn/pull/27308#discussion_r192265008) [](ancestors = 192265008)

AlekseyTs replied:
> For consistency with other DynamicMemberReferences, I'll make this the DynamicType

That would make sense only if there is a separate dynamic invocation for this node, i.e. two dynamic invocations in IL rather than one. Please check IL, and if there is only one dynamic invocation, the type should be null because this is not a value from compiler's point of view

---
In reply to: [192269007](https://github.com/dotnet/roslyn/pull/27308#discussion_r192269007) [](ancestors = 192269007,192265008)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -765,7 +767,15 @@ private IDynamicInvocationOperation CreateBoundDynamicCollectionElementInitializ
+                                                                                                               memberName: "Add",
+                                                                                                               implicitReceiver.Syntax,
+                                                                                                               implicitReceiver.Type,
+                                                                                                               implicitReceiver.ConstantValue,
```
>implicitReceiver.ConstantValue [](start = 111, length = 30)

Similarly, should be default, I think.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1067,7 +1077,11 @@ private IMemberInitializerOperation CreateBoundMemberInitializerOperation(BoundA
 
-            Lazy<IOperation> target = new Lazy<IOperation>(() => Create(boundAssignmentOperator.Left));
+            Lazy<IOperation> target = new Lazy<IOperation>(() =>
+                boundAssignmentOperator.Left.Kind == BoundKind.ObjectInitializerMember && boundAssignmentOperator.Right.Kind == BoundKind.CollectionInitializerExpression
```
>boundAssignmentOperator.Right.Kind == BoundKind.CollectionInitializerExpression [](start = 90, length = 79)

What if the right is regular object initializer?

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1067,7 +1077,11 @@ private IMemberInitializerOperation CreateBoundMemberInitializerOperation(BoundA
-            Lazy<IOperation> target = new Lazy<IOperation>(() => Create(boundAssignmentOperator.Left));
+            Lazy<IOperation> target = new Lazy<IOperation>(() =>
+                boundAssignmentOperator.Left.Kind == BoundKind.ObjectInitializerMember && boundAssignmentOperator.Right.Kind == BoundKind.CollectionInitializerExpression
+                    ? CreateBoundObjectInitializerMemberOperation((BoundObjectInitializerMember)boundAssignmentOperator.Left, inCollectionInitializer: true)
```
>inCollectionInitializer [](start = 126, length = 23)

The name feels somewhat misleading, the target isn't really inside a collection initializer, it is being initialized with a collection initializer.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -841,6 +841,47 @@ static void Main()
+      Initializers(1):
+          IDynamicInvocationOperation (OperationKind.DynamicInvocation, Type: dynamic, IsImplicit) (Syntax: 'd1')
+            Expression: 
+              IDynamicMemberReferenceOperation (Member Name: ""Add"", Containing Type: null) (OperationKind.DynamicMemberReference, Type: C1, IsImplicit) (Syntax: 'C1')
```
>Type: C1 [](start = 132, length = 8)

This doesn't look right.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -615,7 +615,7 @@ private IObjectOrCollectionInitializerOperation CreateBoundCollectionInitializer
         }
 
-        private IOperation CreateBoundObjectInitializerMemberOperation(BoundObjectInitializerMember boundObjectInitializerMember)
+        private IOperation CreateBoundObjectInitializerMemberOperation(BoundObjectInitializerMember boundObjectInitializerMember, bool inCollectionInitializer = false)
```
>BoundObjectInitializerMember [](start = 71, length = 28)

Do we ever expect to run into this node outside of the context of an object initializer? If not, then we should probably make new parameter required and simply remove the case handling this node in the switch at the top of this file. We should also make sure the right thing happens when someone requests IOperation node for the corresponding syntax directly, rather than getting to it through the parent.

333fred replied:
Yes, we'll see it for top-level initializers.

---
In reply to: [192269046](https://github.com/dotnet/roslyn/pull/27308#discussion_r192269046) [](ancestors = 192269046)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1067,7 +1077,11 @@ private IMemberInitializerOperation CreateBoundMemberInitializerOperation(BoundA
-            Lazy<IOperation> target = new Lazy<IOperation>(() => Create(boundAssignmentOperator.Left));
+            Lazy<IOperation> target = new Lazy<IOperation>(() =>
+                boundAssignmentOperator.Left.Kind == BoundKind.ObjectInitializerMember && boundAssignmentOperator.Right.Kind == BoundKind.CollectionInitializerExpression
+                    ? CreateBoundObjectInitializerMemberOperation((BoundObjectInitializerMember)boundAssignmentOperator.Left, inCollectionInitializer: true)
```
>CreateBoundObjectInitializerMemberOperation((BoundObjectInitializerMember)boundAssignmentOperator.Left, inCollectionInitializer: true) [](start = 22, length = 134)

Is this call bypassing the cache? Is it possible to end up with two distinct IOperation nodes for the target?

333fred replied:
It is bypassing the cache, but because we fetch IOperations by getting the enclosing statement and walking down, we shouldn't end up with 2 nodes here.

---
In reply to: [192269300](https://github.com/dotnet/roslyn/pull/27308#discussion_r192269300) [](ancestors = 192269300)

AlekseyTs commented:
Done with review pass (iteration 1)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [393718591](https://github.com/dotnet/roslyn/pull/27308#issuecomment-393718591) [](ancestors = 393718591)

333fred commented:
@dotnet-bot retest windows_release_unit32_prtest please

333fred commented:
@dotnet-bot retest windows_release_vs-integration_prtest please

333fred commented:
@dotnet/roslyn-compiler I need a second review please.

AlekseyTs commented:
@cston, @dotnet/roslyn-compiler Please review, need second review

AlekseyTs commented:
@mavasani, @dotnet/analyzer-ioperation Please review

agocke commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -765,7 +767,10 @@ private IDynamicInvocationOperation CreateBoundDynamicCollectionElementInitializ
-            return new LazyDynamicInvocationExpression(operation, arguments, argumentNames: ImmutableArray<string>.Empty, argumentRefKinds: ImmutableArray<RefKind>.Empty, _semanticModel, syntax, type, constantValue, isImplicit);
+            BoundImplicitReceiver implicitReceiver = boundCollectionElementInitializer.ImplicitReceiver;
+            Lazy<IOperation> addReference = new Lazy<IOperation>(() =>
+                CreateBoundDynamicMemberAccessOperation(implicitReceiver, typeArgumentsOpt: ImmutableArray<TypeSymbol>.Empty, memberName: "Add", implicitReceiver.Syntax, type: null, value: default, isImplicit: true));
```
This is a very long line. Consider splitting it.

333fred replied:
Will do in a follow up. Doesn't look long on a 4K monitor :).

---
In reply to: [192826248](https://github.com/dotnet/roslyn/pull/27308#discussion_r192826248) [](ancestors = 192826248)

agocke commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -9277,5 +9195,457 @@ class C2 : IEnumerable<int>
+    static int GetInt(int i) => i;
+    public C2 this[int i] { get => null; }
+}
+class C2 : IEnumerable<int>
```
Why does this implement `IEnumerable<int>`? I think I'm missing something about this test case.

333fred replied:
Because I'm using a nested collection initializer to initialize it.

---
In reply to: [192830340](https://github.com/dotnet/roslyn/pull/27308#discussion_r192830340) [](ancestors = 192830340)

agocke replied:
Got it. Thanks.

333fred commented:
@agocke I think I addressed your comments.

333fred commented:
@dotnet-bot retest windows_debug_spanish_unit32_prtest please

333fred commented:
Test failures were from the test systems disconnecting, fyi @dotnet/roslyn-infrastructure. Merging as this was just a rebase, and previous pr builds have passed.
